src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp

```diff
@@ -0,0 +1,122 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/GrTagsDeclarations.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace PUP {
+class er;
+}  // namespace PUP
+
+namespace EinsteinSolutions {
+
+/*!
+ * \ingroup EinsteinSolutionsGroup
+ * \brief Kerr black hole in Kerr-Schild coordinates
+ *
+ * \details
+ * The metric is \f$g_{\mu,\nu} = \eta_{\mu\nu} + 2 H l_\mu l_\nu\f$,
+ * where \f$\eta_{\mu\nu}\f$ is the Minkowski metric, \f$H\f$ is a scalar
+ * function, and \f$l_\mu\f$ is the outgoing null vector.
+ * \f$H\f$ and \f$l_\mu\f$ are known functions of the coordinates
+ * and of the mass and spin vector.
+ *
+ * The following are input file options that can be specified:
+ * Mass (default: 1.), Center (default: {0,0,0}), Spin (default: {0,0,0})
+ */
+class KerrSchild {
+ public:
+  struct Mass {
+    using type = double;
+    static constexpr OptionString help = {"Mass of the black hole"};
+    static type default_value() { return 1.; }
+    static type lower_bound() { return 0.; }
+  };
+  struct Spin {
+    using type = std::array<double, 3>;
+    static constexpr OptionString help = {"The [x,y,z] spin of the black hole"};
+    static type default_value() { return {{0., 0., 0.}}; }
+  };
+  struct Center {
+    using type = std::array<double, 3>;
+    static constexpr OptionString help = {
+        "The [x,y,z] center of the black hole"};
+    static type default_value() { return {{0., 0., 0.}}; }
+  };
+  using options = tmpl::list<Mass, Spin, Center>;
+  using type = KerrSchild;  // This class is its own options creator.
```

Comment:
In Test_KerrSchild.cpp, `opts.get<EinsteinSolutions::KerrSchild>()` fails if I leave this out.

In the example I see in the code, there is a helper class that contains a `type` type alias, and then one uses the helper class as the template parameter to opts.get.  But here it seems easier to me to just have KerrSchild itself be the helper class, so you avoid creating an extra class.

Reply:
It's an interesting thought, and I've been going back and forth on whether I like this idea.  It has the definite advantage of requiring less code, and that could be cleaned up even more by defaulting the `type` parameter to the option struct class.  (That couldn't create weird types because plain option structs are not constructible in the option parsing sense.)

I've decided I'm going to discourage it, though.  The two classes have different purposes and different restrictions.  The class `help` should give a reasonably detailed description of the class (which in this case happens to be short), but the option struct `help` has severe length limits (56 characters) and I worry this will encourage people to try to fit their class `help` into those limits.  More importantly, the messages have different purposes: the class `help` says what the class *is*, but the option `help` says what it is *for*.  There may be several option structs that create the same class.  An entry in an input file described as "KerrSchild: Black hole in Kerr-Schild coordinates" gives no hint as to what it is for and probably should read "BackgroundKerrSchild: The background metric" or "InitialData: Initial data parameters" or similar.

(Digression: I'm uncertain about whether we should include the class in the option name.  The class is clear from the help text, but not from just reading the input file.)

For testing purposes I've been meaning to write something like `test_factory_creation` for non-factory things.  I'll see if I can get that in today.

Reply:
Ok, I am now using a separate option parsing struct.  I hadn't thought about multiple option parsing structs creating the same class.